### PR TITLE
Improve Development Experience + Update Socket.IO Dependency

### PR DESCRIPTION
**Pull request by clurdish**
There are a few things this PR accomplishes, mostly to make contributions easier and safer.
1. Updating Socket.IO dependency to match server-side (as of the below PR) 
    - https://github.com/samuraiexx/roll_together_backend/pull/5
2. Add source maps for development builds of the plugin
3. Enable strict mode in TypeScript 
    - And fix errors as a result of turning that on
4. Add types to remote updates to allow TypeScript guards to do their thing
5. Add `window` type for background for better type checking across windows
6. Add `env.json` file to automatically include environment variables
    - In this case it allowed testing both the server and client locally which just changing a config setting
7. Conform single quotes to double quotes across TypeScript files

### CODE REVIEW

**Comment by clurdish** [comment]
@samuraiexx @hsnavarro Let me know if you have any feedback on this. Happy to discuss any of the changes.

File: src/content_script.ts [hunk 1]
```diff
@@ -187,19 +184,19 @@ function handleRemoteUpdate({ roomState, roomProgress }: { roomState: States, ro
   }
 }
 
-function handleBackgroundMessage(args) {
+function handleBackgroundMessage(args: BackgroundMessage) {
```
Comment (samuraiexx) [inline src/content_script.ts:187]: Can you rename args to backgroundMessage? I used a sloppy name hahah

**Comment by samuraiexx** [comment]
Looks great! Thank you very much for those changes. I'll also change the server to update the socket.io version and improve the path a little bit (since I need to support old versions I can't just change the server version because of compatibility between socket.io versions). By the way, is there a specific feature of the new version that you think it's worth using?
I will wait for @hsnavarro to review before merging but looks amazing!

File: src/common.ts [hunk 2]
```diff
@@ -33,6 +46,35 @@ export enum BackgroundMessageTypes {
...
+  animeName: string;
+  begin: number;
+  end: number;
+  episode: number;
+  id: number;
+}
+
+export type BackgroundMessage = RemoteUpdateBackgroundMessage | RoomConnectionBackgroundMessage | SkipMarksBackgroundMessage;
+
+interface BackgroundMessageBase {
```
[the remark this answers is not in the document]
Comment (clurdish) [reply src/common.ts:59]: Yep, TypeScript doesn't care if you export the base interface. Plus using `BackgroundMessageBase` on its own doesn't mean anything.

File: src/common.ts [hunk 3]
```diff
@@ -112,3 +154,27 @@ export function getIntroFeatureState(): Promise<boolean> {
     );
   });
 }
+
+export function enumKeys<O extends object, K extends keyof O = keyof O>(obj: O): K[] {
+  return Object.keys(obj).filter(k => Number.isNaN(+k)) as K[];
```
[the remark this answers is not in the document]
Comment (clurdish) [reply src/common.ts:159]: Not particularly, but I will add comments explaining how/why this function works. That way it's not just magic, which I assume is your concern here.

**Comment by clurdish** [comment]
I have submitted a PR to the server branch for updating Socket.IO and adding TypeScript support to bring the server in line with the client. See: https://github.com/samuraiexx/roll_together_backend/pull/5

The good news with the server side is that it is backwards compatible with older client versions. So deploying the server update first wouldn't disrupt existing clients, no need to worry about deploying at the exact same time.

In terms of updating to the newer version, I don't see any specific features yet. But I think there are a couple reasons to update
1. Better TypeScript support. The newer versions include their own type definitions, which are much more reliable
1. Stay up to date with an actively maintained major version

**Comment by clurdish** [comment]
@hsnavarro Any feedback on this from your end?

**Comment by hsnavarro** [comment]
> @hsnavarro Any feedback on this from your end?

Hey, I already told @samuraiexx to approve it, but he probably forgot. I'll remind him again.

**Pull request merged**